src/System.Memory/src/System/SpanExtensions.cs

```diff
@@ -274,5 +274,104 @@ public static void CopyTo<T>(this T[] array, Span<T> destination)
         }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool Overlaps<T>(this Span<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            return Overlaps((ReadOnlySpan<T>)first, second, out elementOffset);
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                return (uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                       (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>());
+            }
+            else
+            {
+                return (ulong)byteOffset < (ulong)((long)first.Length * Unsafe.SizeOf<T>()) ||
+                       (ulong)byteOffset > (ulong)-((long)second.Length * Unsafe.SizeOf<T>());
+            }
+        }
+
+        /// <summary>
+        /// Determines whether two sequences overlap and outputs the element offset.
+        /// </summary>
+        public static bool Overlaps<T>(this ReadOnlySpan<T> first, ReadOnlySpan<T> second, out int elementOffset)
+        {
+            // note: the calculations below overflow if second.IsEmpty;
+            //       the case second.IsEmpty MUST be handled separately
+            if (first.IsEmpty || second.IsEmpty)
+            {
+                elementOffset = 0;
+                return false;
+            }
+
+            IntPtr byteOffset = Unsafe.ByteOffset(
+                ref first.DangerousGetPinnableReference(),
+                ref second.DangerousGetPinnableReference());
+
+            if (Unsafe.SizeOf<IntPtr>() == sizeof(int))
+            {
+                if ((uint)byteOffset < (uint)(first.Length * Unsafe.SizeOf<T>()) ||
+                    (uint)byteOffset > (uint)-(second.Length * Unsafe.SizeOf<T>()))
+                {
+                    if ((int)byteOffset % Unsafe.SizeOf<T>() != 0)
```

Comment:
@ektrah avoid the `%` by using the `elementOffset` computed below, and use subtract instead with this, if the remaining is not 0 no alignment, and we avoid the modulo. In my mind this also reads better.

Reply:
Of course, you have to multiply too, but multiplying is fast.

Reply:
I believe the % will be optimized to a bitwise-And if sizeof(T) is a power-of-two or entirely optimized away if sizeof(T) is 1, which should account for the overwhelming majority of span uses.

Reply:
@ektrah true but the same could be said for divide and multiply, and this would be more optimal for non-power-of-two sizes e.g. 3-byte `Rgb`. But ok.